### PR TITLE
master <-- JCSP-419: DC,Cloud: Project issue navigator

### PR DESCRIPTION
**Pull request by NataliiaShavshyna**
https://jdog.jira-dev.com/browse/JCSP-419

### CODE REVIEW

File: custom-vu/src/main/kotlin/jces1209/vu/CloudProjectIssueNavigatorScenario.kt [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import com.atlassian.performance.tools.jiraactions.api.action.Action
+import com.atlassian.performance.tools.jiraactions.api.measure.ActionMeter
+import com.atlassian.performance.tools.jiraactions.api.memories.UserMemory
+import com.atlassian.performance.tools.jiraactions.api.scenario.Scenario
+import jces1209.vu.action.LogInWithAtlassianId
+import jces1209.vu.action.ProjectIssueNavigatorAction
+import jces1209.vu.page.issue.CloudProjectIssueNavigatorPage
+import org.openqa.selenium.TakesScreenshot
+
+class CloudProjectIssueNavigatorScenario : Scenario {
```
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/CloudProjectIssueNavigatorScenario.kt]: `Scenario` defines the entire VU behavior.
The goal is to add an `ProjectIssueNavigatorAction` to existing `Scenario`s rather than replacing the entire scenario.

- [x] add an action to existing `Scenario` instead of replacing scenario

File: custom-vu/src/main/kotlin/jces1209/vu/action/ProjectIssueNavigatorAction.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    private val meter: ActionMeter,
+    private val projectMemory: ProjectMemory,
+    private val projectIssueNavigatorPage: AbstractProjectIssueNavigatorPage,
+    private val numberOfProject: String
+
+) : Action {
+    private val logger: Logger = LogManager.getLogger(this::class.java)
+
+    override fun run() {
+        val project = projectMemory.recall()
```
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/action/ProjectIssueNavigatorAction.kt]: It's unused. Is the action doing what it intended?

- [x] use the project memory or remove it

File: custom-vu/src/main/kotlin/jces1209/vu/page/issue/AbstractProjectIssueNavigatorPage.kt [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package jces1209.vu.page.issue
+
+interface AbstractProjectIssueNavigatorPage {
```
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/page/issue/AbstractProjectIssueNavigatorPage.kt]: We don't have to call out the "abstract" part. This will do just fine:
```suggestion
interface ProjectIssueNavigatorPage {
```
`AbstractIssuePage` is called that way only because there already exists an `IssuePage`: `com.atlassian.performance.tools.jiraactions.api.page.IssuePage`.

File: custom-vu/src/main/kotlin/jces1209/vu/page/issue/AbstractProjectIssueNavigatorPage.kt [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package jces1209.vu.page.issue
```
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/page/issue/AbstractProjectIssueNavigatorPage.kt]:
```suggestion
package jces1209.vu.page.project
```

File: custom-vu/src/main/kotlin/jces1209/vu/page/issue/CloudProjectIssueNavigatorPage.kt [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import jces1209.vu.wait
+import org.openqa.selenium.By
+import org.openqa.selenium.WebDriver
+import org.openqa.selenium.support.ui.ExpectedConditions
+import java.time.Duration
+
+class CloudProjectIssueNavigatorPage(
+    private val driver: WebDriver
+) : AbstractProjectIssueNavigatorPage {
+    override fun openProjectByIndex(numberOfProject: String) {
```
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/page/issue/CloudProjectIssueNavigatorPage.kt]: `number` and `String` have incompatible meanings.
Let's use the Kotlin/Java typesystem to our advantage.
```suggestion
    override fun openProjectByIndex(index: Int) {
```

File: custom-vu/src/main/kotlin/jces1209/vu/page/issue/CloudProjectIssueNavigatorPage.kt [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+                    ExpectedConditions.or(
+                        ExpectedConditions.presenceOfElementLocated(By.id("jira-issue-header")),
+                        ExpectedConditions.presenceOfElementLocated(By.id("key-val"))
+                    ),
+                    ExpectedConditions.or(
+                        ExpectedConditions.presenceOfElementLocated(By.id("new-issue-body-container")),
+                        ExpectedConditions.presenceOfElementLocated(By.className("issue-body-content"))
+                    )
+                ),
+                ExpectedConditions.presenceOfElementLocated(By.className("no-results-hint")) // TODO is it too optimistic like in SearchServerFilter.waitForIssueNavigator ?
```
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/page/issue/CloudProjectIssueNavigatorPage.kt]: This looks familiar. Let's create/reuse an abstraction.

- [x] reuse issue-navigator code rather than copy-paste

File: src/test/kotlin/JiraPerformanceComparisonIT.kt [hunk 7]
```diff
@@ -12,6 +12,7 @@ import com.atlassian.performance.tools.virtualusers.api.config.VirtualUserTarget
 import com.atlassian.performance.tools.workspace.api.RootWorkspace
 import com.atlassian.performance.tools.workspace.api.TestWorkspace
 import jces1209.BenchmarkQuality
+import jces1209.QuickAndDirty
```
Comment (dagguh) [inline src/test/kotlin/JiraPerformanceComparisonIT.kt]: Unused import

File: src/test/kotlin/JiraPerformanceComparisonProjectIssueNavigatorIT.kt [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+import jces1209.vu.DcProjectIssueNavigatorScenario
+import org.apache.logging.log4j.core.config.ConfigurationFactory
+import org.junit.Test
+import java.nio.file.Paths
+import java.time.Duration
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors.newCachedThreadPool
+
+class JiraPerformanceComparisonProjectIssueNavigatorIT {
```
Comment (dagguh) [inline src/test/kotlin/JiraPerformanceComparisonProjectIssueNavigatorIT.kt]: I'd  reuse `JiraPerformanceComparisonIT`. Create a whole new test for every action will be hard to use.

**Review by dagguh — CHANGES_REQUESTED** [review]

**Comment by dagguh** [comment]
PS. Can't find the JCSP-419 ticket. Is the LH project not suitable?

File: custom-vu/src/main/kotlin/jces1209/vu/JiraDcScenario.kt [hunk 9]
```diff
@@ -61,6 +59,13 @@ class JiraDcScenario : Scenario {
                 jira = jira,
                 meter = meter,
                 projectMemory = similarities.projectMemory
+            ),
+            projectIssueNavigatorAction = ProjectIssueNavigatorAction(
+                jira = jira,
+                meter = meter,
+                projectMemory = similarities.projectMemory,
+                projectIssueNavigatorPage = DcProjectIssueNavigatorPage(jira.driver),
```
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/JiraDcScenario.kt]: Seems we could reuse `ProjectIssueNavigatorAction` within `ScenarioSimilarities` and pass only the `ProjectIssueNavigatorPage`.

File: custom-vu/src/main/kotlin/jces1209/vu/ScenarioSimilarities.kt [hunk 10]
```diff
@@ -58,6 +53,7 @@ class ScenarioSimilarities(
             editProbability = 0.00f, // 0.10f if we can mutate data
             commentProbability = 0.00f, // 0.04f if we can mutate data
             linkIssueProbability = 0.00f // 0.10f if we can mutate data
+
```
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/ScenarioSimilarities.kt]: I'd avoid unnecessary diff. They generate unnecessary Git conflicts.

File: custom-vu/src/main/kotlin/jces1209/vu/ScenarioSimilarities.kt [hunk 11]
```diff
@@ -103,7 +101,9 @@ class ScenarioSimilarities(
         browseProjects: Action,
         browseFilters: Action,
         browseBoards: Action,
-        viewBoard: Action
+        viewBoard: Action,
+        projectIssueNavigator: Action
```
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/ScenarioSimilarities.kt]: Note the naming of actions. They're in *imperative mood*, that is - they sound like commands.
```suggestion
        browseProjectIssues: Action
```

File: custom-vu/src/main/kotlin/jces1209/vu/action/ProjectIssueNavigatorAction.kt [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+import com.atlassian.performance.tools.jiraactions.api.BROWSE_PROJECTS
+import com.atlassian.performance.tools.jiraactions.api.WebJira
+import com.atlassian.performance.tools.jiraactions.api.action.Action
+import com.atlassian.performance.tools.jiraactions.api.measure.ActionMeter
+import com.atlassian.performance.tools.jiraactions.api.memories.ProjectMemory
+import jces1209.vu.page.project.ProjectIssueNavigatorPage
+import org.apache.logging.log4j.LogManager
+import org.apache.logging.log4j.Logger
+
+class ProjectIssueNavigatorAction(
```
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/action/ProjectIssueNavigatorAction.kt]: Other actions are named like commands
```suggestion
class BrowseProjectIssues(
```

File: custom-vu/src/main/kotlin/jces1209/vu/action/ProjectIssueNavigatorAction.kt [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+    override fun run() {
+        val project = projectMemory.recall()
+        if (project == null) {
+            logger.debug("Skipping Open project issue navigator...")
+            return
+        }
+        meter.measure(BROWSE_PROJECTS) {
+            jira.goToBrowseProjects(1).getProjects()
+            projectIssueNavigatorPage.openProjectByIndex(numberOfProject)
```
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/action/ProjectIssueNavigatorAction.kt]: Let's use consistent naming
```suggestion
            projectIssueNavigatorPage.openProjectByIndex(projectIndex)
```

File: custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt [hunk 14]
```diff
@@ -2,9 +2,7 @@ package jces1209.vu.page
 
 import org.openqa.selenium.By
 import org.openqa.selenium.WebDriver
-import org.openqa.selenium.support.ui.ExpectedConditions.and
-import org.openqa.selenium.support.ui.ExpectedConditions.or
-import org.openqa.selenium.support.ui.ExpectedConditions.presenceOfElementLocated
+import org.openqa.selenium.support.ui.ExpectedConditions.*
```
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/page/CloudIssueNavigator.kt:5]: There's a lot of these `import` diffs in unrelated files.
It makes the PR harder to read, the Git history harder to read and increases risk of Git conflicts.
Also, if it is generated by your IDE, but not by IDE of other collaborators, it risks back-and-forth changes in every PR. I'd align IDE settings within the team and open a dedicated PR to fix the codestyle in the entire repo at once.

Tools like EditorConfig or ktlint can make this alignment easier.

- [x] less noise in the Git history

File: custom-vu/src/main/kotlin/jces1209/vu/ScenarioSimilarities.kt [hunk 15]
```diff
@@ -45,7 +39,8 @@ class ScenarioSimilarities(
         browseBoardsPage: BrowseBoardsPage,
         createIssue: Action,
         searchWithJql: Action,
-        browseProjects: Action
+        browseProjects: Action,
+        projectIssueNavigatorAction: ProjectIssueNavigatorAction
```
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/ScenarioSimilarities.kt]: We can leave it open-ended like for others:
```suggestion
        projectIssueNavigatorAction: Action
```

File: custom-vu/src/main/kotlin/jces1209/vu/page/project/ProjectIssueNavigatorPage.kt [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+                    ),
+                    ExpectedConditions.or(
+                        ExpectedConditions.presenceOfElementLocated(By.id("new-issue-body-container")),
+                        ExpectedConditions.presenceOfElementLocated(By.className("issue-body-content"))
+                    )
+                ),
+                ExpectedConditions.presenceOfElementLocated(By.className("no-results-hint")) // TODO is it too optimistic like in SearchServerFilter.waitForIssueNavigator ?
+            )
+        )
+    }
```
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/page/project/ProjectIssueNavigatorPage.kt]: Very similar selectors are already done in: `SearchServerFilter` and `SearchCloudFilter` (via `CloudIssueNavigator`). Perhaps we could reuse `CloudIssueNavigator` and extract a `ServerIssueNavigator` and reuse it as well.
PS. Server/DC naming is inconsistent, sorry for that. Feel free to standardize on one or the other.

- [x] reuse (non-project-specific) issue navigator page objects

**Review by dagguh — CHANGES_REQUESTED** [review]

File: custom-vu/src/main/kotlin/jces1209/vu/action/ProjectIssueNavigatorAction.kt (hunk 2 above)
Comment (NataliiaShavshyna) [reply custom-vu/src/main/kotlin/jces1209/vu/action/ProjectIssueNavigatorAction.kt]: remove

Comment (NataliiaShavshyna) [reply custom-vu/src/main/kotlin/jces1209/vu/action/ProjectIssueNavigatorAction.kt]: remove

File: custom-vu/src/main/kotlin/jces1209/vu/page/issue/CloudProjectIssueNavigatorPage.kt (hunk 6 above)
Comment (NataliiaShavshyna) [reply custom-vu/src/main/kotlin/jces1209/vu/page/issue/CloudProjectIssueNavigatorPage.kt]: added new abstract class. Name and refactor IssueNavigator added into technical debt.

File: custom-vu/src/main/kotlin/jces1209/vu/page/project/ProjectIssueNavigatorPage.kt (hunk 16 above)
Comment (NataliiaShavshyna) [reply custom-vu/src/main/kotlin/jces1209/vu/page/project/ProjectIssueNavigatorPage.kt]: added abstraction

File: custom-vu/src/main/kotlin/jces1209/vu/JiraDcScenario.kt (hunk 9 above)
Comment (NataliiaShavshyna) [reply custom-vu/src/main/kotlin/jces1209/vu/JiraDcScenario.kt]: they are different

File: custom-vu/src/main/kotlin/jces1209/vu/ScenarioSimilarities.kt (hunk 10 above)
Comment (NataliiaShavshyna) [reply custom-vu/src/main/kotlin/jces1209/vu/ScenarioSimilarities.kt]: will take into consideration in the future

File: custom-vu/src/main/kotlin/jces1209/vu/JiraDcScenario.kt (hunk 9 above)
Comment (dagguh) [reply custom-vu/src/main/kotlin/jces1209/vu/JiraDcScenario.kt]: You mean `projectIndex = 3` vs `projectIndex = 5`? Why are these different?

File: custom-vu/src/main/kotlin/jces1209/vu/ScenarioSimilarities.kt [hunk 17]
```diff
@@ -92,32 +91,33 @@ class ScenarioSimilarities(
...
         browseFilters: Action,
         browseBoards: Action,
-        viewBoard: Action
+        viewBoard: Action,
+        projectIssueNavigator: Action
     ): List<Action> {
         val exploreData = listOf(browseProjects, browseFilters, browseBoards)
         val spreadOut = mapOf(
             createIssue to 0, // 5 if we can mutate data
-            customizeColums to 30,
```
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/ScenarioSimilarities.kt:113]: Looks like deleted by mistake. This PR basically reverts https://github.com/atlassian/jces-1209/pull/54

- [ ] restore `customizeColums`

File: custom-vu/src/main/kotlin/jces1209/vu/page/CommonIssueNavigator.kt [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+            ),
+            ExpectedConditions.or(
+                ExpectedConditions.presenceOfElementLocated(By.id("jira-issue-header")),
+                ExpectedConditions.presenceOfElementLocated(By.id("key-val"))
+            ),
+            ExpectedConditions.or(
+                ExpectedConditions.presenceOfElementLocated(By.id("new-issue-body-container")),
+                ExpectedConditions.presenceOfElementLocated(By.className("issue-body-content"))
+            ),
+            ExpectedConditions.presenceOfElementLocated(By.className("no-results-hint")) // TODO is it too optimistic like in SearchServerFilter.waitForIssueNavigator ?
```
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/page/CommonIssueNavigator.kt]: These conditions are still copied in `CloudIssueNavigator` (even with the same comment). Let's pick one of them and reuse that. I think it's better to keep `CloudIssueNavigator`, because it reuses `FalliblePage`, doesn't use inheritance, has proper imports and wouldn't generate unnecessary diff.

File: custom-vu/src/main/kotlin/jces1209/vu/page/CommonIssueNavigator.kt [hunk 19]
```diff
@@ -0,0 +1,26 @@
+package jces1209.vu.page
+
+import org.openqa.selenium.By
+import org.openqa.selenium.support.ui.ExpectedCondition
+import org.openqa.selenium.support.ui.ExpectedConditions
+
+abstract class CommonIssueNavigator {
```
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/page/CommonIssueNavigator.kt]: [Composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance)

File: custom-vu/src/main/kotlin/jces1209/vu/page/project/DcProjectIssueNavigatorPage.kt [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+import jces1209.vu.page.CommonIssueNavigator
+import jces1209.vu.wait
+import org.openqa.selenium.By
+import org.openqa.selenium.WebDriver
+import org.openqa.selenium.support.ui.ExpectedConditions
+import java.time.Duration
+
+class DcProjectIssueNavigatorPage(
+    private val driver: WebDriver
+) : ProjectIssueNavigatorPage, CommonIssueNavigator() {
```
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/page/project/DcProjectIssueNavigatorPage.kt]: DC has a different navigator. See `SearchServerFilter.waitForIssueNavigator`.

Perhaps I wasn't clear. The global issue navigator (GIN) and project issue navigator (PIN) embed the same issue navigator component. But they're still different across Cloud and DC.

- [x] Separate Cloud and DC issue navigators
- [ ] test with DC

**Review by dagguh — CHANGES_REQUESTED** [review]

**Review by dagguh — CHANGES_REQUESTED** [review]
Last thing remaining is to keep `customizeColumns` intact

File: custom-vu/src/main/kotlin/jces1209/vu/ScenarioSimilarities.kt (hunk 17 above)
Comment (NataliiaShavshyna) [reply custom-vu/src/main/kotlin/jces1209/vu/ScenarioSimilarities.kt:113]: ill fix this during conflict resolving

File: custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+            or(
+                presenceOfElementLocated(By.cssSelector("ol.issue-list")),
+                presenceOfElementLocated(By.id("issuetable")),
+                presenceOfElementLocated(By.id("issue-content"))
+            ),
+            presenceOfElementLocated(By.id("key-val")),
+            presenceOfElementLocated(By.className("issue-body-content"))
+        )
+    )
+        .cloudErrors()
```
Comment (dagguh) [inline custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt]: `cloudErrors` in DC page object? Perhaps you meant Server errors?

- [ ] detect DC errors

**Review by dagguh — CHANGES_REQUESTED** [review]

File: custom-vu/src/main/kotlin/jces1209/vu/action/ProjectIssueNavigatorAction.kt [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+    private val meter: ActionMeter,
+    private val projectKeyMemory: ProjectMemory
+
+) : Action {
+    private val logger: Logger = LogManager.getLogger(this::class.java)
+
+    override fun run() {
+        val projectKey = projectKeyMemory.recall()
+        if (projectKey == null) {
+            logger.debug("I don't recall any issue keys. Maybe next time I will.")
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/action/ProjectIssueNavigatorAction.kt]:
```suggestion
            logger.debug("I don't recall any project keys. Maybe next time I will.")
```

File: custom-vu/src/main/kotlin/jces1209/vu/action/ProjectIssueNavigatorAction.kt [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+            return
+        }
+        meter.measure(BROWSE_PROJECTS) {
+            jira.goToProjectSummary(projectKey.key)
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/action/ProjectIssueNavigatorAction.kt]: also, it's necessary to add `wait for page load` - FalliblePage should help with it
or use the new page classes

File: custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+
+import org.openqa.selenium.By
+import org.openqa.selenium.WebDriver
+import org.openqa.selenium.support.ui.ExpectedConditions.*
+import java.time.Duration
+
+class DcIssueNavigator(
+    driver: WebDriver
+) {
+    private val page = FalliblePage.Builder(
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt]:
```suggestion
    private val falliblePage = FalliblePage.Builder(
```

File: custom-vu/src/main/kotlin/jces1209/vu/action/ProjectIssueNavigatorAction.kt [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+) : Action {
+    private val logger: Logger = LogManager.getLogger(this::class.java)
+
+    override fun run() {
+        val projectKey = projectKeyMemory.recall()
+        if (projectKey == null) {
+            logger.debug("I don't recall any issue keys. Maybe next time I will.")
+            return
+        }
+        meter.measure(BROWSE_PROJECTS) {
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/action/ProjectIssueNavigatorAction.kt]: Please, create a new record in MeasureType - like `Browse project issues`

**Review by ysemenkov — CHANGES_REQUESTED** [review]

File: custom-vu/src/main/kotlin/jces1209/vu/page/project/ProjectIssueNavigatorPage.kt [hunk 26]
```diff
@@ -0,0 +1,9 @@
+package jces1209.vu.page.project
+
+import org.openqa.selenium.WebDriver
+
+interface ProjectIssueNavigatorPage {
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/project/ProjectIssueNavigatorPage.kt]: This new interface is not being used

File: custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt [hunk 27]
```diff
@@ -0,0 +1,30 @@
+package jces1209.vu.page
+
+import org.openqa.selenium.By
+import org.openqa.selenium.WebDriver
+import org.openqa.selenium.support.ui.ExpectedConditions.*
+import java.time.Duration
+
+class DcIssueNavigator(
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/DcIssueNavigator.kt:8]: It looks like it must have the same interface as CloudIssueNavigator

File: custom-vu/src/main/kotlin/jces1209/vu/page/project/ProjectIssueNavigatorPage.kt [hunk 28]
```diff
@@ -0,0 +1,9 @@
...
+
+    fun openProjectByIndex(index: Int)
+    fun waitForProjectIssueNavigator(driver: WebDriver)
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/project/ProjectIssueNavigatorPage.kt]:
```suggestion
    fun waitForNavigator()
```

File: custom-vu/src/main/kotlin/jces1209/vu/page/project/ProjectIssueNavigatorPage.kt [hunk 29]
```diff
@@ -0,0 +1,9 @@
+package jces1209.vu.page.project
+
+import org.openqa.selenium.WebDriver
+
+interface ProjectIssueNavigatorPage {
+
+    fun openProjectByIndex(index: Int)
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/project/ProjectIssueNavigatorPage.kt]: Project is being loaded from project key from memory
Let's make it so
```suggestion
    fun openProject(projectKey: String)
```

**Review by ysemenkov — CHANGES_REQUESTED** [review]

File: custom-vu/src/main/kotlin/jces1209/vu/page/project/DcProjectNavigatorPage.kt [hunk 30]
```diff
@@ -0,0 +1,23 @@
...
+import jces1209.vu.wait
+import org.openqa.selenium.By
+import org.openqa.selenium.WebDriver
+import org.openqa.selenium.support.ui.ExpectedConditions
+
+class DcProjectNavigatorPage(
+    private val driver: WebDriver
+) : ProjectNavigatorPage {
+    override fun openProject(projectKey: Int) {
+        driver
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/project/DcProjectNavigatorPage.kt:14]: There is no navigation to any page

Note: the same for Cloud

File: custom-vu/src/main/kotlin/jces1209/vu/page/project/CloudProjectNavigatorPage.kt [hunk 31]
```diff
@@ -0,0 +1,25 @@
...
+
+
+class CloudProjectNavigatorPage(
+    private val driver: WebDriver
+) : ProjectNavigatorPage {
+
+    override fun openProject(projectKey: Int) {
+        driver
+            .wait(ExpectedConditions
+                .elementToBeClickable(By.xpath("(//*[@data-testid='Content']//a)[$projectKey]")))
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/project/CloudProjectNavigatorPage.kt]: `[@data-testid='Content']` is the total page container
use this one `data-test-id="global-pages.directories.directory-base.content.table.container"`

File: custom-vu/src/main/kotlin/jces1209/vu/action/BrowseProjectIssues.kt [hunk 32]
```diff
@@ -0,0 +1,32 @@
...
+    private val logger: Logger = LogManager.getLogger(this::class.java)
+
+    override fun run() {
+        val projectKey = projectKeyMemory.recall()
+        if (projectKey == null) {
+            logger.debug("I don't recall any projects keys. Maybe next time I will.")
+            return
+        }
+        meter.measure(BROWSE_PROJECT_ISSUES) {
+            browseProjectPage.openProject((projectKey.key).toInt())
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/action/BrowseProjectIssues.kt]: Project.key doesn't contain int
it's `Key` column value

**Review by ysemenkov — CHANGES_REQUESTED** [review]

File: custom-vu/src/main/kotlin/jces1209/vu/page/project/CloudProjectNavigatorPage.kt [hunk 33]
```diff
@@ -0,0 +1,29 @@
...
+
+    override fun openProject(projectKey: String) {
+        driver.navigate().to("/projects/")
+        driver
+            .wait(ExpectedConditions
+                .elementToBeClickable(
+                    By.xpath(
+                        "//*[@data-test-id='global-pages.directories.directory-base.content.table.container']" +
+                            "//*[text()='$projectKey']")))
+            .click()
```
Comment (ysemenkov) [inline custom-vu/src/main/kotlin/jces1209/vu/page/project/CloudProjectNavigatorPage.kt:22]: Is it working? `Key` column doesn't contain link

**Review by ysemenkov — APPROVED** [review]

**Comment by ysemenkov** [comment]
@NataliiaShavshyna don't forget to squash commits

**Comment by NataliiaShavshyna** [comment]
Closed by second approve
